toolboxes/feel/feelmodels/fluid/fluidmechanicscreate.cpp

```diff
@@ -899,7 +899,7 @@ NullSpace<double> getNullSpace( SpaceType const& space, mpl::int_<3> /**/ )
 
 FLUIDMECHANICS_CLASS_TEMPLATE_DECLARATIONS
 void
-FLUIDMECHANICS_CLASS_TEMPLATE_TYPE::init( bool buildModelAlgebraicFactory )
+FLUIDMECHANICS_CLASS_TEMPLATE_TYPE::init( bool buildModelAlgebraicFactory, bool buildBlockVector )
```

Comment:
@metivett why it's necessary to have this arg buildBlockVector? It changes the standard behavior of a toolbox that I assumed: after init, the block vector must be built, it's very important in multiphysics toolboxes. Also, it doesn't take a lot of memory (it's a just a vector and it's very fast)
Moreover, it breaks the python interface.

Reply:
@vincentchabannes It is currently required by the `MultiFluid` initialisation, because of the possible coupling between the `FluidMechanics` and the `LevelSet` classes: in the case of an imposed surface inextensibility constraint, we need the `LevelSet`'s function-space to build the Lagrange-multiplier bloc in the algebraic vector, and we also need the `FluidMechanics` velocity space for the level-set advection.
Actually, we already discussed this kind of design when removing the two-step initialisation behaviour (with a `build` and an `init` member), which provided much more flexibility for the coupling between toolboxes, at the cost of maintainability. Maybe we can find another more robust initialisation approach ?

Reply:
You can build the block vector in the init of fluid and upgrade this one in the init of multifluid. It’s very important to avoid intrusivity for maintains thé code. You canot change Api, the impact is big now and I want all toolboxes similar. I would also prefer to not have buildVector as virtual.

Reply:
It still seems that the call to `M_blockVectorSolution.buildVector( this->backend() );` needs all the blocks to be properly constructed,  no ?
If I use the approach you suggest (build a $(U,p)$ vector in `FluidMechanics`, then "upgrade" it to add the required $L$ Lagrange-multiplier field), I can't really update the vector, I need to rebuild it completely (all the more since the `resize` function, which actually wraps the STL::vector::resize, is not guaranteed to leave the already constructed parts untouched)... Even if this is not the most costly part (and by far, I agree), this does somehow look a bit flawed, suggesting that our API is not fully robust.

Reply:
@metivett I have moved `M_blockVectorSolution.buildVector( this->backend() );` into the initAlgebraicFactory ( Everything that depends on the monolithic formulation is built here).
Also, I did a change which probably breaks your code, you need to call `FM->startTimeStep()` at the beginning of the time loop else the bdf will not be started. I did that for implement Theta time stepping and a better management of initial solution

Reply:
@vincentchabannes Ok, I will look into this ! Do you agree to the changes I did to remove intrusion into `FluidMechanics` ?

Reply:
Completly agree

Reply:
@vincentchabannes Regarding the changes you made to `FluidMechanics`, maybe we should rename the `initAlgebraicFactory` to something more explicit if this function is supposed to build all the monolithic algebraic data (maybe something like `initAlgebraicData` ?). Moreover, we should remove the `buildBlockVector`, which now only forwards the call to `initBlockVector`...
And @vincentchabannes @prudhomm, regarding the time stepping strategy, are we planning any refactoring soon ? The interface between generic time-stepper, and the strong dependance on `BDF` API is somehow clumsy, and the toolboxes should abstract themselves from the details of the time-stepping used.

Reply:
I am agree with the renaming, I didn't want to break your codes.

The bdf object is still used with theta scheme just by simplicity. A lot of time schemes (as theta scheme) are writed in the form u^{n+1} = u^n + \Delta t * F, so we can rewrite as  (u^{n+1} - u^n )/Delta t = F and we retrive the bdf order 1. This form allow to have a F which not depends of time step and avoid to have switch between bdf and other time stepper. I am agree that use of bdf should be take with caution. My next step is to implement the fractional theta step scheme (or more generaly multiple step scheme) and we will see what it's the best way to do that.

Reply:
I don't exactly get your point regarding the time-stepping... Actually, it seems that if you write most of the schemes as (u^{n+1} - u^n )/Delta t = F, then F should depend on the scheme (e.g. on \theta in the theta-scheme). And there are lots of schemes which can't be written under this form (except if you put everything in F :wink:), like for example the Crank-Nicolson, which involves (u^{n+1} - u^n )/Delta t - 1/2 F^{n+1} = 1/2 F^{n}.
I think that the time-stepping classes should reflect this difference and provide a clear interface using some `timeStepBase` class, and then have multiple inheritance chains. For the particular case of the theta scheme, I think it should have its own class, allowing thereby the Crank-Nicolson special case theta=1/2.

Reply:
We support now Crank-nickolson, it's a theta scheme with theta=0.5. F should depends of time n+1 but also of time n, n-1,... I do a framework for manage exactly this kind of scheme.

Reply:
My works is actualy done for only for theta scheme, but it's now very easy to add others time schemes as adams-moulton

Reply:
Ok, I have merged `develop` into `feature/ls` and fixed the issues related to the new `startTimeStep` of `FluidMechanics` in f0a34c6.